### ISSUE TITLE
Suggestion: make sure that `project.theme` is always an object

### ISSUE THREAD

**Issue by alexvoss** [mkdocs/mkdocs#3436]
Dear all,

working on Material for MkDocs, @squidfunk and I came across the problem that `project.theme` is sometimes an object and sometimes a string. The latter happens when the user has not configured a theme and the default MkDocs scheme is applied. In that case, `project.theme == "mkdocs"`. Another situation is where people write, e.g., `theme: material` in their `mkdocs.yml`. 

It would be good to standardize this to be always an object with the theme name in `project.theme.name`. This could either be done under the hood, leaving both options open for people's `mkdocs.yml` or the documentation could also be changed so that people are instructed to use an object in their YAML. Currently different pages show different versions: 

*  https://www.mkdocs.org/getting-started/#theming-our-documentation (theme)
*  https://www.mkdocs.org/user-guide/choosing-your-theme/ (theme.name)

IMHO it would be better to always advise people to use the object version as themes tend to have options that need to be configured. In that case the user has to switch to the object version anyway, which might even be confusing to people who were not born speaking YAML.

Cheers, Alex

**Comment by oprypin** [comment]
Hi.
As far as I can tell, the issue you're describing does not actually happen under any normal conditions.

It is easy to try - add `print(config.theme)` in any hook on a basic mkdocs site, and it will be a `Theme` object with fields such as `name` and others.
It will be so in all of these cases:
* `theme: {name: mkdocs}`
* `theme: mkdocs`
* `# unspecified`

It is already done under the hood and it shouldn't be a surprise. If the user-specified dictionary was directly stored in the config, you'd not be able to write `config.theme.name` anyway, it'd have to be `config.theme['name']`.

---

I can believe you that something like this happens in your case, but you're not providing any reproduction instructions whatsoever, so I'm unable to advise.

* I do actually know of an edge case where raw config dictionaries sneak into the config like you're describing. It can happen when one config option fails validation but other code still keeps executing until all config items finish their validation. Could that be also what happens in your case?

   I fixed a specific symptom of this in #3351 but haven't fixed the root problem.

**Comment by alexvoss** [comment]
Thanks for the swift feedback. "It is already done under the hood" - I think I know where I went wrong. I should have realized when I wrote `project` instead of `config`. 

@squidfunk, I think we need to implement the equivalent logic that already goes on under the hood in MkDocs in the projects plugin. Will have a look to see what other transformations take place when the `mkdocs.yml` is being read - if any.

I do think it would be good to tweak the documentation of MkDocs to guide people to using he object format by default, see the two links posted.
